File: sop-sdk/sdk-python/request/BaseRequest.py
#!/usr/bin/python
# -*- coding: UTF-8 -*-
from response.BaseResponse import BaseResponse


class BaseRequest:
    """请求类的父类"""

    biz_model = None
    """请求参数"""

    files = None
    """上传文件"""

    def __init__(self):
        pass

    def get_method(self):
        """返回接口名

        :return: 返回接口名
        :rtype: str
        """
        raise Exception('未实现BaseRequest.get_method()方法')

    def get_version(self):
        """返回接口版本号

        :return: 返回版本号，如：1.0
        :rtype: str
        """
        raise Exception('未实现BaseRequest.get_version()方法')

    def get_request_type(self):
        """返回请求类型

        :return: 返回RequestType类实例
        :rtype:  common.RequestType
        """
        raise Exception('未实现BaseRequest.get_request_type()方法')

    def parse_response(self, response_dict):
        response_data = response_dict.get('error_response')
        if response_data is None:
            data_name = self.get_method().replace('.', '_') + '_response'
            response_data = response_dict.get(data_name)
        base_response = BaseResponse(response_data)
        base_response.request_id = response_dict.get('request_id')
        base_response.code = response_data.get('code')
        base_response.msg = response_data.get('msg')
        base_response.sub_code = response_data.get('sub_code')
        base_response.sub_msg = response_data.get('sub_msg')
        return base_response

File: sop-sdk/sdk-nodejs/common/BaseRequest.js
const isArray = require('isarray');
/**
 * 请求类父类
 */
module.exports = class BaseRequest {
    constructor() {
        this.bizModel = {};

        this.files = undefined;

        // 用于文件上传时强制转换成POST_FILE请求
        this.__forceRequestType__ = undefined;

        this.checkOverride();
    }

    /**
     * 校验子类是否已重写相关方法
     * */
    checkOverride() {
        try {
            this.getMethod();
            this.getVersion();
            this.getRequestType();
        } catch (error) {
            throw error;
        }
    }

    setBizModel(biz = {}) {
        this.bizModel = biz;
        return this;
    }

    setFiles(files) {
        this.files = files;
        return this;
    }

    addFile(name, path) {
        if (name && path) {
            if (!isArray(this.files)) {
                this.files = [];
            }
            this.files.push({name, path});
        }
        return this;
    }

    /**
     * 返回接口名称
     */
    getMethod() {
        throw `未实现BaseRequest类getMethod()方法`;
    }

    /**
     * 返回版本号
     */
    getVersion() {
        throw '未实现BaseRequest类getVersion()方法';
    }

    /**
     * 返回请求类型，使用RequestType.js
     */
    getRequestType() {
        throw '未实现BaseRequest类getRequestType()方法';
    }

    setForceRequestType(type) {
        this.__forceRequestType__ = type;
        return this;
    }

    getRealRequestType() {
        return this.__forceRequestType__ || this.getRequestType();
    }

    /**
     * 解析返回结果，子类可以覆盖实现
     * @param responseData 服务器返回内容
     * @returns 返回结果
     */
    parseResponse(responseData) {
        let data = responseData['error_response'];
        if (!data) {
            const dataNodeName = this.getMethod().replace(/\./g, '_') + '_response';
            data = responseData[dataNodeName];
        }
        return data;
    }
};
